pkg/cri/server/image_pull.go

```diff
@@ -786,3 +794,31 @@ func (rt *pullRequestReporterRoundTripper) RoundTrip(req *http.Request) (*http.R
 	}
 	return resp, err
 }
+
+// Given that runtime information is not passed from PullImageRequest, we depend on an experimental annotation
+// passed from pod sandbox config to get the runtimeHandler. The annotation key is specified in configuration.
+// Once we know the runtime, try to override default snapshotter if it is set for this runtime.
+// See https://github.com/containerd/containerd/issues/6657
+func (c *criService) snapshotterFromPodSandboxConfig(s *runtime.PodSandboxConfig) (string, error) {
+	snapshotter := c.config.ContainerdConfig.Snapshotter
+	runtimeAnnotationKey := c.config.ContainerdConfig.PullImageRuntimeAnnotationKey
+	if runtimeAnnotationKey == "" {
+		return snapshotter, nil
+	}
+
+	if s == nil || s.Annotations == nil {
+		return snapshotter, nil
+	}
+
+	runtimeHandler, ok := s.Annotations[runtimeAnnotationKey]
+	if ok {
+		ociRuntime, err := c.getSandboxRuntime(s, runtimeHandler)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to get sandbox runtime for %s", runtimeHandler)
+		}
+
+		snapshotter = c.runtimeSnapshotter(ociRuntime)
```

Comment:
let's be loud here adding an info level log entry about this being a prototype and that we are switching to the runtime configured snapshotter

Reply:
Done, added more logging when there's runtime snapshotter overrides